app/models/minion.rb

```diff
@@ -19,27 +19,36 @@ class NonExistingMinion < StandardError; end
   # Example:
   #   Minion.assign_roles(
   #     roles: {
-  #       master: 1,
+  #       master: [1],
   #       minion: [2, 3]
   #     },
   #     default_role: :dns
   #   )
   def self.assign_roles!(roles: {}, default_role: :minion)
-    requested_master = roles[:master].to_i
-    requested_minions =
-      roles[:minion] || Minion.unassigned_role.pluck(:id) - [requested_master]
-    requested_minions = requested_minions.map(&:to_i)
+    requested_masters = select_masters(roles)
+    requested_minions = select_minions(roles, requested_masters)
 
-    if !requested_master.blank? && !Minion.exists?(id: requested_master)
-      raise NonExistingMinion, "Failed to process non existing minion id: #{requested_master}"
+    requested_masters.each do |requested_master|
+      unless Minion.exists?(id: requested_master)
+        raise NonExistingMinion, "Failed to process non existing master id: #{requested_master}"
```

Comment:
Maybe it makes sense to simply have an exception class called `NonExistingNode`, which covers both masters and workers ? Just saying because it feels a bit weird to raise a `NonExistingMinion` for a master 😉